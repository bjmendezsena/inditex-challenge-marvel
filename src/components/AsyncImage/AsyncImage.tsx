import { ComponentProps, useState } from 'react';
import NextImage from 'next/image';
import { Spinner } from '@/components';
import clsx from 'clsx';

export type AsncImageProps = {
  isLoading?: boolean;
} & ComponentProps<typeof NextImage>;

export const AsyncImage = ({ isLoading, ...props }: AsncImageProps) => {
  const [isLoadingImage, setIsLoadingImage] = useState(true);
  const handleLoadedImage = () => {
    setIsLoadingImage(false);
  };

  return (
    <>
      {(isLoadingImage || isLoading) && (
        <Spinner data-testid="async-image-spinner" />
      )}
      <NextImage
        onLoad={handleLoadedImage}
        className={clsx({
          'display-none': isLoadingImage,
        })}
        {...props}
      />
    </>
  );
};
