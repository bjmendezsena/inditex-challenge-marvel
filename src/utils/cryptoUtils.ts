import Crypto from "crypto-js";

export function createHashMd5(text: string): string {
  return Crypto.MD5(text).toString();
}
