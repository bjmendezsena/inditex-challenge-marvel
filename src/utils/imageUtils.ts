import { Thumbnail } from '@/types';

export function buildThumbnailImage(thumbnail: Thumbnail) {
  const { extension, path } = thumbnail;
  return `${path.replace('http', 'https')}.${extension}`;
}
