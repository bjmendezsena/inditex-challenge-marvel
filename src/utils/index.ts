export * from "./cryptoUtils";
export * from "./urls";
export * from "./format";
