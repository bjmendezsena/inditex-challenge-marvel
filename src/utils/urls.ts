export interface Routes {
  [path: string]:
    | string
    | Routes
    | (() => string)
    | ((...args: any[]) => string);
}

export interface UrlBuilder {
  <R extends Routes>(path: string, routes: R): R & { toString(): string };
}

export const urlBuilder: UrlBuilder = (base, routes) => {
  const mappedRoutes: Routes = {
    toString() {
      return base;
    },
  };

  Object.keys(routes).forEach((route) => {
    const url = routes[route];
    if (typeof url === "function" && route === "toString") {
      // Check if is method
      mappedRoutes.toString = function () {
        return base + routes.toString();
      };
    } else if (typeof url === "object") {
      mappedRoutes[route] = urlBuilder(base, url);
    } else if (typeof url === "string") {
      if (url === "") {
        mappedRoutes[route] = mappedRoutes.toString();
      } else {
        const path =
          url.indexOf("/") === 0
            ? url
            : [base, url].join("/").replace("//", "/");
        mappedRoutes[route] = path;
      }
    } else {
      throw new TypeError(
        `"${route}" is not valid. A routes object can only contain a string, an object or the "toString" method as values.`
      );
    }
  });

  return mappedRoutes as typeof routes;
};
