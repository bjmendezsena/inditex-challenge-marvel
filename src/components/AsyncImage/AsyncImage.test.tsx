import _merge from 'lodash/merge';
import { render } from '@/tests/utils';
import { AsyncImage, AsncImageProps } from './AsyncImage';

const defaultProps: AsncImageProps = {
  src: 'https://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784',
  alt: 'test',
  width: 100,
  height: 100,
};

describe(`<${AsyncImage.name} />`, () => {
  const getComponent = (props: Partial<AsncImageProps> = {}) =>
    render(<AsyncImage {..._merge(defaultProps, props)} />);
  it('should render without crashing', () => {
    const { asFragment } = getComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with isLoading', () => {
    const { asFragment, getByTestId } = getComponent({ isLoading: true });

    expect(asFragment()).toMatchSnapshot();

    expect(getByTestId('async-image-spinner')).toBeInTheDocument();
  });
});
