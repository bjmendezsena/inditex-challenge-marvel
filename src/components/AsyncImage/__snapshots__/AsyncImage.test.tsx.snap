// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AsyncImage /> > should render with isLoading 1`] = `
<DocumentFragment>
  <div
    class="load"
    data-testid="async-image-spinner"
  >
    <div
      class="load__icon-wrap"
    >
      <svg
        class="load__icon"
      >
        <path
          d="M12,4V2A10,10 0 0,0 2,12H4A8,8 0 0,1 12,4Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <img
    alt="test"
    class="display-none"
    data-nimg="1"
    decoding="async"
    height="100"
    loading="lazy"
    src="/_next/image?url=https%3A%2F%2Fi.annihil.us%2Fu%2Fprod%2Fmarvel%2Fi%2Fmg%2Fc%2Fe0%2F535fecbbb9784&w=256&q=75"
    srcset="/_next/image?url=https%3A%2F%2Fi.annihil.us%2Fu%2Fprod%2Fmarvel%2Fi%2Fmg%2Fc%2Fe0%2F535fecbbb9784&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.annihil.us%2Fu%2Fprod%2Fmarvel%2Fi%2Fmg%2Fc%2Fe0%2F535fecbbb9784&w=256&q=75 2x"
    style="color: transparent;"
    width="100"
  />
</DocumentFragment>
`;

exports[`<AsyncImage /> > should render without crashing 1`] = `
<DocumentFragment>
  <div
    class="load"
    data-testid="async-image-spinner"
  >
    <div
      class="load__icon-wrap"
    >
      <svg
        class="load__icon"
      >
        <path
          d="M12,4V2A10,10 0 0,0 2,12H4A8,8 0 0,1 12,4Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <img
    alt="test"
    class="display-none"
    data-nimg="1"
    decoding="async"
    height="100"
    loading="lazy"
    src="/_next/image?url=https%3A%2F%2Fi.annihil.us%2Fu%2Fprod%2Fmarvel%2Fi%2Fmg%2Fc%2Fe0%2F535fecbbb9784&w=256&q=75"
    srcset="/_next/image?url=https%3A%2F%2Fi.annihil.us%2Fu%2Fprod%2Fmarvel%2Fi%2Fmg%2Fc%2Fe0%2F535fecbbb9784&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.annihil.us%2Fu%2Fprod%2Fmarvel%2Fi%2Fmg%2Fc%2Fe0%2F535fecbbb9784&w=256&q=75 2x"
    style="color: transparent;"
    width="100"
  />
</DocumentFragment>
`;
